### DISCUSSION TITLE
RFC: Persistent Cache Storage

### DISCUSSION THREAD

**Discussion by LingyuCoder** [web-infra-dev/rspack#8646] [RFC]
Feature Name: Design the storage module of persistent cache
Start Date: 2024/12/09

# Summary
Persistent cache storage module design:
1. Read cached data from the file system before starting a phase of the build process
2. Write the resulting cached data after a phase of the build process is completed

The storage module does not care about how the data is generated and consumed (what occassion modules do) or the serialization/deserialization of the data (what serializers do), but only focuses on the organization and reading/writing of the data in the file system.

# Motivation

Cached data consists of KV pairs, and isolated cached data is generated during the `make`, `seal`, and other phases. Therefore, the storage system needs to have the following capabilities:
1. A storage instance is divided into multiple storage areas (named `scope`) for storing caches generated/consumed at each phase of the build process.
  1. A scope can be read on demand before caching consumption in the corresponding phase to prevent reading too much data at once from blocking the main process
  2. A scope can be written after the building process enters the idle phase, and atomicity should be ensured as much as possible to prevent the situation where only part of the cache is written, making the overall cache unavailable
2. A scope maps to multiple files on file system and needs to control the size of the files
  1. A single file should not be too large (such as storing all KVs in one file), which makes it difficult to concurrently read and write on demand
  2. A single file should not be too small (such as one file per KV), resulting in too many files and too many IO operations.
  3. When the cache is modified in incremental building, the writing of files should be minimized as much as possible. The unmodified part should not be written.
3. There should be basic validation for cached files to prevent data from being unavailable in the following scenarios
  1. Stored files are modified, deleted, etc.
  2. Storage configuration changes

# Glossary

1. `Storage`: A storage instance mounted on the compiler, so it will cross multiple compilation under dev mode
2. `Scope`: A storage area distinguished by name, will be used in different building phases
3. `Pack`: A storage block which maps to a stored file on file system, containing multiple KV data items.
4. `Meta`: A metafile used to manage storage packs of a scope, used to read and verify storage blocks of this scope.

# User Guide

For the upper-level persistent cache modules, a storage instance exposes the following interfaces:

```Rust
#[async_trait::async_trait]
pub trait Storage: std::fmt::Debug + Sync + Send {
  async fn load(&self, scope: &'static str) -> Result<Vec<(Arc<Vec<u8>>, Arc<Vec<u8>>)>>;
  fn set(&self, scope: &'static str, key: Vec<u8>, value: Vec<u8>);
  fn remove(&self, scope: &'static str, key: &Vec<u8>);
  fn trigger_save(&self) -> Result<Receiver<Result<()>>>;
}
```

- `load()` : Asynchronously reads all data items under a Scope
- `set()`: Add/modify specified data items under scope, will not be written immediately, will be uniformly processed in the last idle
- `remove()` : Remove the specified data item under the scope, it will not be written immediately, and it will be uniformly processed in the last idle
- `trigger_save()` : notify storage to write all data, which will be processed in an independent thread, and notify the write result through the oneshot channel

```rust
async fn compiler() {
  let storage = Storage::new();
  
  let compilation = Compilation::new();
  // make phase
  let raw_make_cache = storage.load("make").await;
  let make_cache = deserialize(&raw_make_cache).await;
  for item in serialize(make().await) {
    storage.set/* or remove */("make", item.key, item.value);
  }
  
  // seal phase
  let raw_seal_cache = storage.load("seal").await;
  let seal_cache = deserialize(&raw_make_cache).await;
  for item in serialize(make().await) {
    storage.set/* or remove */("seal", item.key, item.value);
  }
  
  // after process assets
  let rx = storage.trigger_save();
  
  // build: wait for saving
  let res = rx.recv().await; 
  match res {
    Err(e) => compilation.addDiagnostic(e),
    _ => ()
  }


  // watch: just print the error
  tokio::spawn(async move || {
    let res = rx.recv().await;
    match res {
      Err(e) => infrastructure_logger.warn(format!("{}", e)),
      _ => ()
    }
  });
  
  compilation_done().await;  
  
  let next_compilation = Compilation::clone(&compilation);
  
  for item in serialize(incremental_make().await) {
    storage.set/* or remove */("seal", item.key, item.value);
  }
  
  let rx = storage.trigger_save();
  // ...
}
```

It should be noted that the waiting for idle channels can be different according to the mode:
1. `compiler.build()`: should wait and add storage error to diagnostics
2. `compiler.watch()`: no need to wait, the data will be updated to scopes when saved, and the write task will use the queue to execute in order. If the storage fails, the error from the channel needs to be print into `stderr`.
# Detailed design

## Modules
Include several parts:
1. Storage API: provides API for upper layer, almost no logic
2. Manager: management module used to schedule reading, updating, and writing of scopes. Internally includes:
  1. Scopes: Storage area for storing data in memory
  2. Updates: used to temporarily store the data items which will be merged into scopes when idle
3. Task Queue: a thread write queue, used to asynchronously execute write operations to the scopes when idle
4. Strategy: define the logic of validating, reading, updating, and writing on the scope, will split a scope to multiple packs
5. File System: interacting with intermediate file system
6. Meta File/Pack File: the final data files

![image](https://github.com/user-attachments/assets/f377e3eb-6f5a-4bdf-adc4-4e5cd26da3ed)

## Interaction
The APIs are handled as follows:
1. `get_all()`：
  1. First, try to read the meta data of the scope and the keys at the header of pack files from the file system, and verify them. The validation includes the configuration object, expiration time, file size, last modify time, keys hash, etc., and will not verify the complete storage content.
  2. If the validation passes, all contents of the stored file are read and stored in the scope, and the reference is returned for upper-level consumption
2. set/remove：
  1. Temporarily store the item in manager without any processing
3. idle：
  1. Update the temporary items to the scopes, and inform the task queue to write when idle.
  2. In the task queue, take out all scopes for writing, and return ownership after writing is completed. And notify the result of writing through the channel

After the idle call, the latest data can be obtained through the `load()`. However, since the asynchronous write task will acquire ownership of scopes, it will be locked during the write process to prevent the `load()` reading empty values at this time.

```mermaid
sequenceDiagram
activate Compilation
loop every phase
Compilation ->> Manager: load()
activate Manager

Manager -->> FileSystem: read meta
deactivate Manager
activate FileSystem

FileSystem -->> Manager: meta & pack keys
deactivate FileSystem
activate Manager

Manager ->> Manager: validate
Manager -->> FileSystem: read contents
deactivate Manager
activate FileSystem

FileSystem -->> Manager: pack contents
deactivate FileSystem
activate Manager

Manager ->> Compilation: cache items
deactivate Manager

Compilation ->> Compilation: start make phase
activate Compilation

Compilation ->> Manager: set()/remove()
activate Manager
deactivate Manager

Compilation ->> Manager: set()/remove()
activate Manager
deactivate Manager

deactivate Compilation
end

Compilation ->> Manager: trigger_save()
activate Manager
Manager ->> Manager: update scopes
Manager ->> TaskQueue: scopes ownership
activate TaskQueue
Manager ->> Compilation: receiver
deactivate Manager

Compilation ->> Compilation: done()
activate Compilation

TaskQueue ->> FileSystem: write pack files
activate FileSystem
FileSystem ->> TaskQueue: file metadatas
deactivate FileSystem
TaskQueue ->> TaskQueue: create scope meta
deactivate Compilation
TaskQueue ->> FileSystem: write meta file
activate FileSystem
FileSystem ->> TaskQueue: write end
deactivate FileSystem
TaskQueue ->> Manager: scopes ownership
activate Manager

deactivate Manager
TaskQueue -> Compilation: async write result
deactivate TaskQueue

deactivate Compilation

```

## Strategy
Since usually when a rspack project builds, the vast majority of modules come from the `node_modules`, these modules will not change frequently, only a small number of user modules will be frequently modified, in the dev mode this situation is particularly obvious. Therefore, consider using `bucket`s to incrementally write within a certain range:
1. First, the data items are bucketed according to the hash of the key, and the total number of buckets can be configured (like 10 or 20).
2. For all data in the bucket, it is partitioned to packs, and the maximum size of a pack can be configured (like 500kb).
3. After all pack files under bucketing are written, a meta file is generated, which includes the current configuration items of the scope and the final file name and file hash (calculated from keys + mtime + size).

### Validation
1. Read the meta file to load the configuration and the time of writing the storage
2. Compare the current configuration and expiration time, if it does not match, the cache is not available
3. Read the metadata of pack files, obtain mtime and size, and read the keys part of the header of the pack file and calculate the hash.
4. Compared with the hash in the meta file, if it does not match, the cache file has been modified and the cache will not be available
5. Read the whole body of pack files and get the data items
6. If the read data item corresponds to keys, the cache is available

```mermaid
flowchart TB
Validate --> ReadMeta[read meta file]
ReadMeta --> ValidateMeta{options match?}
ValidateMeta -- changed or expired --> Invalid[invalid]
ValidateMeta -- passes --> ReadKeys[read packs & pack keys]
ReadKeys --> ValidateHash{hashes match?}
ValidateHash -- not match --> Invalid
ValidateHash -- passes --> ReadContents[read pack contents]
ReadContents --> ValidateContents{contents match?}
ValidateContents -- length not match --> Invalid
ValidateContents -- passes --> Valid
```

## Updating

Set 80% of the maximum size of a pack as stable size limit:
1. If a stable pack has no data items modified/deleted, its file does not need to be operated on, and the structure in memory will be directly reused
2. If there are no data items modified/deleted in an unstable pack, the pack is also considered dirty due to the possibility of adding new data items
3. If any data item in a pack is modified/deleted, the pack will be dirty
4. For dirty packs, it is believed that they will be modified, that is, the files will be rewritten, so the data in their memory will be moved out and their files will be deleted
5. Merge the data items took from the dirty pack with the newly added/modified/deleted data items to form a new list of data items, and fill the pack one by one according to size until a pack is not filled enough
6. A single data item that exceeds a stable size limit will generate an independent package, regardless of its size

```mermaid
flowchart TD
OldPacks["take(scope.packs)"] --> GetItemMapping[get item to pack mapping]
GetItemMapping --> CheckPackDirty[find out dirty packs]
UpdateItems["take(update[scope_name])"] -- removed/updated items --> CheckPackDirty
CheckPackDirty -- items in dirty packs --> NewItems[generate new items list]
UpdateItems -- insert/updated items --> NewItems
CheckPackDirty -- unchanged packs --> UnchangedPacks[unchanged stable packs]
NewItems -- filled new packs --> CreatedPacks[created new packs]
UnchangedPacks --> FinalPacks[final packs]
CreatedPacks --> FinalPacks
```

## Exception handling
Any errors that occur during the caching process should not block the main process, and in the worst case, only cause the cache to be unavailable. Therefore, errors generated during the process should be regarded as warnings.
1. While calling `load()`, if the validation fails or reads abnormally, it will fallback to the cache is not available, and there will be an empty scope in memory
2. While writing scopes, the files that need to be changed will be written to the temp directory first. If it fails, an error will be sent through the channel, but the data in the memory scope is still available. It can be updated normally and processed at the next write.
3. After all scopes are written to the temp directory, the files in the temp directory will be moved to the cache directory to ensure atomicity as much as possible. Of course, if an error occurs during the movement, it may cause damage to the cache files.
4. After the move is completed, all files in the directory will be detected based on the scope object in the current memory, and irrelevant files will be removed.

# Unresolved Questions & TODOs

1. There is still a lot of room for optimization in the current strategy
  1. Refer to some optimization ideas of webpack
  2. For stable packs, if only few KVs are changed and it is still in a stable state after the change, local changes in the file can be considered
  3. A single large package will currently exceed the limit volume and need to be split
  4. For incremental scenarios, special treatment should be given. Currently, in order to update the cache, a complete copy is kept in memory, which consumes a lot of memory. In the future, we will consider releasing the cache in memory and only reading the affected pack to generate incremental updates in the incremental case
  5. Implement more strategies for different scenarios
2. Exception handling is rough, and exception prompts still need to be optimized.
  1. Provide detailed failure reasons when the validation fails
  2. Provide the cause of the exception and the files related when an exception occurs
3. Other performance issues
  1. At present, when processing data writing, ownership transfer has actually been done, but because cloning is required for upstream consumption, `Arc` is still needed. Perhaps there is a better way
4. Need more test cases and benchmarks

**Comment by jerrykingxyz** [comment]
Feature Name: Persistent Cache Storage V2
Start Date: 2026/03/17

# Summary

Design the storage module for persistent cache, providing the ability to read and write KV data persistently.

The storage module does not concern itself with how data is generated or consumed, nor with serialization/deserialization. It focuses solely on organizing and reading/writing KV data across different storage backends.

# Glossary

- `Storage`: A storage instance mounted on the compiler, persisting across multiple compilations in watch mode
- `Scope`: A virtual storage region used by upper layers, identified by name; different build phases use different scopes; in filesystem storage, each scope maps to one bucket
- `DB`: The underlying database for filesystem storage, designed specifically for persistent cache, managing reads and writes across multiple buckets
- `Bucket`: A data partition within DB, organizing KV data internally through pack files
- `Pack`: A single storage file, corresponding to one `.pack` file on disk, containing a set of KV entries

# User Guide

## Storage trait

The storage instance exposes the following interface to the persistent cache layer above:

```rust
#[async_trait::async_trait]
pub trait Storage: std::fmt::Debug + Sync + Send {
  /// Asynchronously loads all entries under a scope and validates data integrity
  async fn load(&self, scope: &'static str) -> Result<Vec<(Vec<u8>, Vec<u8>)>>;
  /// Inserts or updates an entry under a scope; buffered in memory, not written immediately
  fn set(&self, scope: &'static str, key: Vec<u8>, value: Vec<u8>);
  /// Removes an entry under a scope; buffered in memory, not written immediately
  fn remove(&self, scope: &'static str, key: &[u8]);
  /// Flushes all buffered changes to the storage backend asynchronously;
  /// returns a `Receiver` the caller can await for the result
  fn trigger_save(&self) -> Result<Receiver<Result<()>>>;
  /// Clears all in-memory state and deletes cached data on disk
  async fn reset(&self);
  /// Returns the list of all existing scope names
  async fn scopes(&self) -> Result<Vec<String>>;
}

pub type ArcStorage = Arc<dyn Storage>;
```

## MemoryStorage

A pure in-memory implementation that stores all data in a `HashMap`. Data does not survive process exit and is intended for unit testing.

```rust
let storage = MemoryStorage::default();
```

## FileSystemStorage

A persistent implementation backed by the filesystem, storing KV data on disk using the pack file format.

```rust
pub struct FileSystemOptions {
  /// Root directory for the cache; all versions are stored in subdirectories under this path
  pub directory: Utf8PathBuf,
  /// Version identifier, corresponding to a subdirectory under `directory`
  pub version: String,
  /// Maximum size of a single pack file in bytes; data overflows into a new pack when exceeded
  pub max_pack_size: usize,
  /// Old version directories not accessed within this duration (seconds) are deleted; `0` means never expire
  pub expire: u64,
  /// Filesystem implementation; use `NativeFileSystem` in production
  pub fs: Arc<dyn IntermediateFileSystem>,
}

let storage = FileSystemStorage::new(FileSystemOptions {
  directory: "/path/to/cache".into(),
  version: "1.0.0".into(),
  max_pack_size: 4 * 1024 * 1024,
  expire: 7 * 24 * 3600,
  fs: Arc::new(NativeFileSystem),
});
```

## Detailed design

This section focuses on the internal implementation of `FileSystemStorage`. `MemoryStorage` is straightforward and not covered here.

### File Layout

`FileSystemStorage` organizes all cached data under the `directory` path. Each `version` maps to an independent subdirectory for isolation; different versions do not interfere with each other, and stale version directories are cleaned up automatically after expiry.

Within each version directory, each scope maps to a bucket subdirectory containing pack files and a `_meta` index file. On write, all new files are first staged under `.temp/` and then renamed to their final location upon commit.

```
{directory}/                        ← FileSystemOptions.directory, cache root
├── _meta                           ← Last-access timestamps for each version, used for expiry cleanup
│                                      Format: one line per version: "{version} {unix_timestamp}"
└── {version}/                      ← FileSystemOptions.version, current version subdirectory
    ├── .temp/                      ← Write staging area, deleted after commit;
    │   │                              if present at startup, indicates the previous process crashed
    │   ├── state.lock              ← Records PID + process name of the current lock holder,
    │   │                              used to determine whether the previous writer is still running
    │   ├── commit.lock             ← WAL file listing all files to add/remove in this commit;
    │   │                              format: [ADD] / [REMOVE] sections, one relative path per line
    │   └── {bucket}/               ← Staging location for new files of each bucket
    │       ├── _meta
    │       └── *.pack
    └── {bucket}/                   ← Bucket subdirectory
        ├── _meta                   ← Bucket metadata index, recording a summary for each pack;
        │                              first line: pack ID allocator state (list of used IDs),
        │                              subsequent lines: "{pack_id} {content_hash} {bloom_filter}"
        ├── 0.pack                  ← Hot pack: always in memory, written on every save
        ├── 1.pack                  ← Cold pack: immutable after write, only rewritten if it contains a modified key
        └── N.pack
```

Pack files store KV data in an append-only format: each record begins with a line `"{key_len} {value_len}"` followed by the raw bytes of the key and value. Integrity metadata (`content_hash` and `bloom_filter`) is not stored in the pack file itself; instead, it is saved as a `PackIndex` in the bucket's `_meta` file and verified by recomputing the hash on read.

### Code Architecture

With the file layout in mind, the code structure follows naturally. The relationships between modules are shown below:

<img width="2762" height="884" alt="image" src="https://github.com/user-attachments/assets/7c7081bc-5ab2-4346-b449-5edb884fce55" />

- **FileSystemStorage**: Implements the `Storage` trait. Buffers `set`/`remove` changes in memory and hands them off to DB on `trigger_save()`, returning a `Receiver` immediately without blocking the caller
- **DB**: Maintains one Bucket per scope (lazily initialized); forwards read requests to the corresponding Bucket and enqueues write tasks to TaskQueue, coordinating with Transaction for commit
- **TaskQueue**: A single background tokio task that executes write tasks sequentially, preventing concurrent writes across multiple `trigger_save()` calls
- **Transaction**: Implements two-phase commit; provides Buckets with a staging filesystem pointing to `.temp/`; renames files to their final location after all writes complete; checks for crash recovery at the start of each task
- **Bucket**: Manages all pack files under a single scope; implements the hot/cold pack separation strategy; reads all packs in parallel with integrity validation; uses bloom filters to identify cold packs that need rewriting, then writes in parallel
- **Pack**: Handles reading and writing of a single `.pack` file and KV serialization; produces a `PackIndex` (containing `content_hash` and `BloomFilter`) on each write
- **Meta**: Reads and writes the `_meta` index file; maintains the `PackIndex` list for all packs

### Write Strategy (Hot/Cold Pack Separation)

In a typical Rspack project, the majority of modules come from `node_modules` and rarely change; only a small amount of user code changes frequently, especially in watch mode. Rewriting all pack files on every save would cause severe write amplification. The hot/cold pack separation strategy addresses this: the hot pack absorbs frequent changes while cold packs remain stable once written.

- Each bucket has exactly one **hot pack** (`0.pack`), always kept in memory and written on every save
- When the hot pack exceeds `max_pack_size`, the current contents are flushed to a **cold pack** (`1.pack`, `2.pack`, …)
- Cold packs are immutable after being written; on the next save, only cold packs that contain a modified key are reloaded and rewritten — all others remain untouched on disk

<img width="2754" height="602" alt="image" src="https://github.com/user-attachments/assets/c7373d66-93f0-4850-bff0-d4619497d3de" />

### Crash Safety (Two-Phase Commit)

If a process crashes during a write, the cache must be either fully intact or cleanly invalid on the next startup — partial writes that corrupt data are not acceptable. This guarantee is provided entirely by `Transaction` and is transparent to `Bucket` and `DB`.

`Transaction` splits a write into two phases:

- **Prepare phase**: Provides Buckets with a staging filesystem pointing to `.temp/`; all new files are written there while the original files in the root directory remain untouched
- **Commit phase**: Writes the list of files to add/remove into `commit.lock` (WAL), then performs the renames and deletions

If the process crashes during the prepare phase, the staged content is simply discarded on the next startup. If it crashes during the commit phase, the remaining operations can be replayed from `commit.lock`.

### Version Expiry

As toolchains or configurations are upgraded, the `version` string changes. Stale version directories would accumulate on disk indefinitely without cleanup. After each `trigger_save()`, storage triggers a version expiry check; to avoid extra I/O on every build, the check is rate-limited by a timestamp.

### Key Changes

This refactor preserves the public API. Changes are confined to internal implementation.

1. **Strategy pattern → cohesive modules**  
   The previous design used a Strategy pattern where `SplitPackStrategy` simultaneously implemented `RootStrategy`, `ScopeStrategy`, and `PackWriteStrategy`, concentrating all read, write, validation, and file commit logic in a single struct. The abstraction layer has been removed; responsibilities are now distributed across `DB`, `Bucket`, `Pack`, `Meta`, `Transaction`, and `TaskQueue`.

2. **Bucket grouping layer removed**  
   Each scope previously used `bucket_size` (default 10) subdirectories to hash-distribute keys, resulting in a three-level directory structure. This layer has been removed; each scope now maps directly to a flat directory of pack files.

3. **Bloom filter for write-path optimization**  
   Previously, every update required parsing all pack files to locate affected keys. Now, the `BloomFilter` stored in each pack's `PackIndex` (persisted in `_meta`) allows the write path to quickly identify which cold packs need rewriting, significantly reducing file I/O and memory usage.
